File: libs/export-shapes-modal/src/main/kotlin/mono/export/ExportShapesHelper.kt
package mono.export

import mono.graphics.bitmap.MonoBitmap
import mono.graphics.board.Highlight
import mono.graphics.board.MonoBoard
import mono.graphics.geo.Rect
import mono.shape.shape.AbstractShape
import mono.shape.shape.Group

/**
 * A helper class for exporting selected shapes.
 */
class ExportShapesHelper(
    private val getBitmap: (AbstractShape) -> MonoBitmap?,
    private val setClipboardText: (String) -> Unit
) {

    fun exportText(shapes: List<AbstractShape>, isModalRequired: Boolean) {
        if (shapes.isEmpty()) {
            return
        }

        val left = shapes.minOf { it.bound.left }
        val right = shapes.maxOf { it.bound.right }
        val top = shapes.minOf { it.bound.top }
        val bottom = shapes.maxOf { it.bound.bottom }
        val window = Rect.byLTRB(left, top, right, bottom)

        val exportingBoard = MonoBoard().apply { clearAndSetWindow(window) }
        drawShapesOntoExportingBoard(exportingBoard, shapes)

        val text = exportingBoard.toStringInBound(window)
        if (isModalRequired) {
            ExportShapesModal().show(text)
        } else {
            setClipboardText(text)
        }
    }

    private fun drawShapesOntoExportingBoard(board: MonoBoard, shapes: Collection<AbstractShape>) {
        for (shape in shapes) {
            if (shape is Group) {
                drawShapesOntoExportingBoard(board, shape.items)
                continue
            }
            val bitmap = getBitmap(shape) ?: continue
            board.fill(shape.bound.position, bitmap, Highlight.NO)
        }
    }
}

File: monosketch-svelte/src/lib/mono/state-manager/export/export-shapes-helper.ts
/*
 * Copyright (c) 2024, tuanchauict
 */

import { Rect } from "$libs/graphics-geo/rect";
import { MonoBitmap } from "$mono/monobitmap/bitmap/monobitmap";
import { MonoBoard } from "$mono/monobitmap/board";
import { HighlightType } from "$mono/monobitmap/board/pixel";
import type { AbstractShape } from "$mono/shape/shape/abstract-shape";
import { Group } from "$mono/shape/shape/group";
import ExportShapesModalComponent from "$ui/modal/export/ExportShapesModal.svelte";

/**
 * A helper class for exporting selected shapes.
 */
export class ExportShapesHelper {

    constructor(
        private getBitmap: (shape: AbstractShape) => MonoBitmap.Bitmap | null,
        private setClipboardText: (text: string) => void,
    ) {
    }

    exportText(shapes: AbstractShape[], isModalRequired: boolean): void {
        if (shapes.length === 0) {
            return;
        }

        const left = Math.min(...shapes.map(shape => shape.bound.left));
        const right = Math.max(...shapes.map(shape => shape.bound.right));
        const top = Math.min(...shapes.map(shape => shape.bound.top));
        const bottom = Math.max(...shapes.map(shape => shape.bound.bottom));
        const window = Rect.byLTRB(left, top, right, bottom);

        const exportingBoard = new MonoBoard();
        exportingBoard.clearAndSetWindow(window);
        this.drawShapesOntoExportingBoard(exportingBoard, shapes);

        const text = exportingBoard.toStringInBound(window);
        if (isModalRequired) {
            // TODO: Show modal
            ExportShapesModal.show(text);
        } else {
            this.setClipboardText(text);
        }
    }

    private drawShapesOntoExportingBoard(board: MonoBoard, shapes: Iterable<AbstractShape>): void {
        for (const shape of shapes) {
            if (shape instanceof Group) {
                this.drawShapesOntoExportingBoard(board, shape.items);
                continue;
            }
            const bitmap = this.getBitmap(shape);
            if (bitmap) {
                board.fillBitmap(shape.bound.position, bitmap, HighlightType.NO);
            }
        }
    }
}

const ExportShapesModal = {
    show: (text: string) => {
        const targetElement = document.getElementById('export-text-modal');
        if (!targetElement) {
            console.error('Export text modal container not found');
            return;
        }

        // Clear any existing content
        targetElement.innerHTML = '';

        const modal = new ExportShapesModalComponent({
            target: targetElement,
            props: {
                text,
                onDismiss: () => {
                    modal.$destroy();
                    targetElement.innerHTML = '';
                },
            },
        });
    },
}
